wiring/inc/spark_wiring_ble.h

```diff
@@ -0,0 +1,460 @@
+/*
+ * Copyright (c) 2018 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef SPARK_WIRING_BLE_H
+#define SPARK_WIRING_BLE_H
+
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include "spark_wiring_platform.h"
+
+#if Wiring_BLE
+
+#include "system_error.h"
+#include "spark_wiring_string.h"
+#include "spark_wiring_vector.h"
+#include "spark_wiring_flags.h"
+#include "ble_hal.h"
+#include "check.h"
+#include <memory>
+
+using spark::Vector;
+using particle::Flags;
+
+class BleScanResult;
+class BlePeerDevice;
+
+// Forward declaration
+class BleCharacteristicImpl;
+class BleServiceImpl;
+class BleGattServerImpl;
+class BleGattClientImpl;
+class BleBroadcasterImpl;
+class BleObserverImpl;
+class BlePeripheralImpl;
+class BleCentralImpl;
+
+enum class BleUuidType {
+    SHORT = 0,
+    LONG = 1
+};
+
+enum class BleUuidOrder {
+    MSB = 0,
+    LSB = 1
+};
+
+struct BleDeviceRoleType; // Tag type for BLE device role flags
+typedef Flags<BleDeviceRoleType, uint8_t> BleDeviceRoles;
+typedef BleDeviceRoles::FlagType BleDeviceRole;
+namespace ROLE {
```

Comment:
Too generic of a name. Even if put into `particle` namespace, it still pollutes. Consider moving out of this namespace, renaming HAL roles to `HAL_BLE_ROLE_XXX` and using just `BLE_ROLE_XXX` similar to what we do with `SleepOptionFlag`.